core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala

```diff
@@ -0,0 +1,105 @@
+)(implicit
+  protected val executionContext: ExecutionContext
+) {
+
+  def registerGoogleProject(googleProjectReg: GoogleProjectRegistration): Future[GoogleProjectRegistration] =
+    for {
+      existingGoogleProjectReg <- googleProjectRegRepo.getGoogleProjectRegistration(googleProjectReg.googleProjectId)
+      billingProject <- billingRepository.getBillingProject(googleProjectReg.billingProjectId).map {
+        maybeBillingProject =>
+          maybeBillingProject.getOrElse(
+            throw new RawlsExceptionWithErrorReport(
+              errorReport = ErrorReport(
+                StatusCodes.NotFound,
+                "Billing project does not exist or you do not have permission to perform this action."
+              )
+            )
+          )
+      }
+      _ <- samDAO
+        .listUserActionsForResource(SamResourceTypeNames.billingProject, billingProject.projectName.value, ctx)
+        .map { actions =>
+          if (actions.isEmpty || !actions.contains(SamBillingProjectActions.link))
+            throw new RawlsExceptionWithErrorReport(
+              errorReport =
+                ErrorReport(StatusCodes.NotFound,
+                            "Billing project does not exist or you do not have permission to perform this action."
+                )
+            )
+        }
+      canLinkGoogleProject <- samDAO
+        .userHasAction(SamResourceTypeNames.googleProject,
+                       googleProjectReg.googleProjectId.value,
+                       SamGoogleProjectActions.link,
+                       ctx
+        )
+      _ = if (!canLinkGoogleProject)
+        throw new RawlsExceptionWithErrorReport(
+          errorReport = ErrorReport(StatusCodes.Forbidden, "You do not have permission to perform this action.")
+        )
+      result <- existingGoogleProjectReg match {
+        case Some(existing) if existing.billingProjectId == googleProjectReg.billingProjectId =>
+          Future.successful(existing)
+        case Some(existing) if existing.billingProjectId != googleProjectReg.billingProjectId =>
+          throw new RawlsExceptionWithErrorReport(
+            errorReport = ErrorReport(StatusCodes.Conflict,
+                                      "This google project id is already registered with a different billing project."
+            )
+          )
+        case _ =>
+          val updatedGoogleProjectReg = googleProjectReg.copy(billingAccount = billingProject.billingAccount)
+          for {
+            _ <- googleServicesDAO.setBillingAccountName(googleProjectReg.googleProjectId,
+                                                         billingProject.billingAccount.get,
+                                                         ctx.toTracingContext
+            )
+            registeredGoogleProject <- googleProjectRegRepo.registerGoogleProject(updatedGoogleProjectReg)
+          } yield registeredGoogleProject
+      }
+    } yield result
```

Comment:
in this for comprehension … where should the transaction boundaries be? As-is, you have a chance of a race condition if it receives multiple requests to register the same project at the same time. Admittedly it's a small chance, but it's definitely there

Reply:
Hm, maybe around this whole service method?  Does scala have a transaction annotation equivalent?

Reply:
The "annotation" would be the `dataSource.inTransaction` which you currently have in the Repository class. The catch of putting it around the whole method is that you'll be holding a MySQL transaction open while making REST API calls to Sam. Ideally the Sam calls (and any other non-MySQL operations) would be outside the transaction.

Reply:
Ugh and I just removed the datasource from this file too.  I suppose it would make sense to wrap it around just the part where we check for existing rows in the table, then do any updates, although that would include sitting around waiting to hear back from Google.

Reply:
How about creating the registration record first and checking for conflict at that point and then setting the billing account in Google after? The registration record would need to be rolled back if the billing account change fails, but you wouldn't need to issue the Google call inside a transaction.

Reply:
Ok, I've rearranged the logic so that the checking for conflict and registration happens first, in a transaction, before updating the billing account in Google.  To avoid weirdness in edge cases around setting the billing account, the `registerGoogleProject` method returns `None` if no action was taken, and Some(createdProject) if it was created.  This lets the service layer know whether to attempt updating the billing account in Google or not, and also informs the status code (200 v 201).